faunadb/errors.py

```diff
@@ -1,63 +1,131 @@
 """Error types that methods in the FaunaDB client throw."""
 
-class FaunaError(Exception):
-  """Base class for all FaunaDB errors."""
-  pass
 
-class InvalidQuery(FaunaError):
-  """Thrown when a query is malformed."""
-  pass
+class InvalidResponse(Exception):
+  """Thrown when the response from the server is unusable."""
+  def __init__(self, description, response_data):
+    self.description = description
+    """Description of the kind of response expected."""
+    self.response_data = response_data
+    """Actual response data. (Type varies.)"""
+    super(InvalidResponse, self).__init__("%s\nGot: %s" % (description, response_data))
 
 
-class InvalidValue(FaunaError):
-  """Thrown when a value cannot be accepted."""
-  def __init__(self, message="The field value is not valid."):
-    super(InvalidValue, self).__init__(message)
-
+#region FaunaHttpError
+class FaunaError(Exception):
+  """
+  Error returned by the FaunaDB server.
+  For documentation of error types, see the `docs <https://faunadb.com/documentation#errors>`__.
+  """
 
-#region HTTPError
-class FaunaHTTPError(FaunaError):
-  """Error in FaunaDB server connection."""
   def __init__(self, response_dict):
-    if "error" in response_dict:
-      self.errors = [response_dict["error"]]
-    else:
-      self.errors = response_dict["errors"]
-    self.reason = response_dict.get("reason", "")
-    self.parameters = response_dict.get("parameters", {})
-    super(FaunaHTTPError, self).__init__(self.reason or self.errors[0])
-
-class BadRequest(FaunaHTTPError):
+    self.errors = [ErrorData.from_dict(error) for error in get_or_invalid(response_dict, "errors")]
+    """List of all :py:class:`ErrorData` objects sent by the server."""
+    super(FaunaError, self).__init__(
+      self.errors[0].description if self.errors else "(empty `errors`)")
+
+  def __repr__(self):
+    return "%s(%s)" % (self.__class__.__name__, [repr(error) for error in self.errors])
```

Comment:
Don't run `self.errors` through `repr()`, as it's adding double quotes: `HttpBadRequest(["ValidationFailed([], [Failure([u'data', u'x'], Value is not unique.)])"])`